csharp/src/ParallelSelectExt.cs

```diff
@@ -0,0 +1,368 @@
+// This file is part of the ArmoniK project
+//
+// Copyright (C) ANEO, 2022-2023.All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using JetBrains.Annotations;
+
+namespace ArmoniK.Utils;
+
+/// <summary>
+///   Options for ParallelSelect and ParallelWait
+/// </summary>
+[PublicAPI]
+public struct ParallelTaskOptions
+{
+  private readonly int parallelismLimit_ = 0;
+
+  /// <summary>Limit the parallelism for ParallelSelect and ParallelWait</summary>
+  public int ParallelismLimit
+  {
+    get
+      => parallelismLimit_ switch
+         {
+           < 0 => int.MaxValue,
+           0   => Environment.ProcessorCount,
+           _   => parallelismLimit_,
+         };
+    init => parallelismLimit_ = value;
+  }
+
+  /// <summary>Cancellation token used for stopping the enumeration</summary>
+  public CancellationToken CancellationToken { get; init; } = default;
+
+  /// <summary>
+  ///   Options for ParallelSelect and ParallelWait.
+  ///   If parallelismLimit is 0, the number of threads is used as the limit.
+  ///   If parallelismLimit is negative, no limit is enforced.
+  /// </summary>
+  /// <param name="parallelismLimit">Limit the parallelism</param>
+  /// <param name="cancellationToken">Cancellation token used for stopping the enumeration</param>
+  public ParallelTaskOptions(int               parallelismLimit,
+                             CancellationToken cancellationToken = default)
+  {
+    ParallelismLimit  = parallelismLimit;
+    CancellationToken = cancellationToken;
+  }
+
+  /// <summary>
+  ///   Options for ParallelSelect and ParallelWait.
+  ///   Parallelism is limited to the number of threads.
+  /// </summary>
+  public ParallelTaskOptions()
+    : this(0)
+  {
+  }
+
+  /// <summary>
+  ///   Options for ParallelSelect and ParallelWait.
+  ///   Parallelism is limited to the number of threads.
+  /// </summary>
+  /// <param name="cancellationToken">Cancellation token used for stopping the enumeration</param>
+  public ParallelTaskOptions(CancellationToken cancellationToken)
```

Comment:
Useless.
One can use named parameter if they want to only provide cancelation token

Reply:
In theory, I agree, but I would much prefer write `.ParallelSelect(new(cancellationToken), x= > ...)` rather than `.ParallelSelect(new(cancellationToken: cancellationToken), x => ...)`.